### ISSUE TITLE
Add support for other SR-IOV devices

### ISSUE THREAD

**Issue by ahalim-intel** [intel/sriov-network-device-plugin#208]
# Add support for other SR-IOV devices

| Status        | Proposed                                             |
:-------------- |:---------------------------------------------------- |
| **Author(s)** | Abdul Halim (@ahalim-intel)                          |
| **Updated**   | 2020-02-14                                           |

## Objective

The objective of this proposal is to extend SR-IOV network device plugin to support other SR-IOV capable devices for example FPGA, Accelerators, SmartNICs etc.

## Motivation

The following PRs were added to support other devices with this plugin which is backed by user demands and requirements.

* https://github.com/intel/sriov-network-device-plugin/pull/196
* https://github.com/intel/sriov-network-device-plugin/pull/125

Since this plugin is already capable of creating multiple Kubelet device plugin endpoints and provides a many SR-IOV specific utility functions, it may also possible to extend it to support these devices without breaking its existing functionality.

The problem with the approach taken with the PRs above, the new devices are also treated same as network devices. This may make it difficult to maintain or introduce device specific changes in future for all devices.

## User Benefit

User could leverage the same device plugin to discover and register other SR-IOV Virtual Function(VF)s.

## Design Proposal

Current implementation of the device plugin focuses on physical SR-IOV networking devices in BareMetal only. The device discovery, resource configurations and selectors are very much network device centric.

This proposal suggests the following changes in current implementation so that handling other type of devices can be done by implementing some shared interfaces for generic SR-IOV device, not limited to network devices only.

1. Extract a generic `PciDevice` interface from `PciNetDevice`. Then each device type will extend this as shown in the example below for `PciNetDevice`.

Current `PciNetDevice` interface:

```go
type PciNetDevice interface {
	GetPFName() string
	GetPfPciAddr() string
	GetVendor() string
	GetDriver() string
	GetDeviceCode() string
	GetPciAddr() string
	GetNetName() string
	IsSriovPF() bool
	GetLinkSpeed() string
	GetLinkType() string
	GetSubClass() string
	GetDeviceSpecs() []*pluginapi.DeviceSpec
	GetEnvVal() string
	GetMounts() []*pluginapi.Mount
	GetAPIDevice() *pluginapi.Device
	GetRdmaSpec() RdmaSpec
	GetVFID() int
	GetDDPProfiles() string
}
```

Change this to: 

```go
// PciDevice provides an interface to get device specific information
type PciDevice interface {
	GetVendor() string
	GetDriver() string
	GetDeviceCode() string
	GetPciAddr() string
	IsSriovPF() bool
	GetSubClass() string
	GetDeviceSpecs() []*pluginapi.DeviceSpec
	GetEnvVal() string
	GetMounts() []*pluginapi.Mount
	GetAPIDevice() *pluginapi.Device
	GetVFID() int
	GetPfPciAddr() string
}

// PciNetDevice extends PciDevice
type PciNetDevice interface {
	PciDevice
	GetPFName() string
	GetNetName() string
	GetLinkSpeed() string
	GetLinkType() string
	GetRdmaSpec() RdmaSpec
    GetDDPProfiles() string
}
```

This will allow the plugin to support other new devices with device specific interface and behaviours. The device specific interface will allow to have device specific `selectors`. For example,

```go
// PciAccelDevice extends PciDevice
type PciAccelDevice interface {
	PciDevice
    SomeFPGAMethod()
}
```

2. Introduce a new user defined resourceConfig parameter `deviceType`. This parameter will be optional and default to `netdevice` for backward compatibility.

```json
{
    "resourceList": [{
            "resourceName": "intel_sriov_netdevice",
            "selectors": {
                "vendors": ["8086"],
                "devices": ["154c", "10ed"],
                "drivers": ["i40evf", "ixgbevf"],
                "pfNames": ["enp0s0f0","enp2s2f1"]
            }
        },
        {
            "resourceName": "intel_sriov_fec",
            "resourcePrefix": "intel.com",
            "deviceType": "accelerator",
            "selectors": {
                "vendors": ["8086"],
                "devices": ["ffff"],
                "drivers": ["vfio-pci"],
                "other_fec_selector": ["values"]
            }
        }
    ]
}
```
3. The SR-IOV plugin manager will keep a map of `PciDevicePool` pool with `deviceType` as key. Each `PciDevicePool` will contain its own list of `PciDevice`. Eeah resource config will create a new kubelet device plugin enpoint serving devices from type of `PciDevicePool`.

4. The `discoverHostDevices()` will discover devices for all supported `deviceType` using device specific methods. 

5. Each `PciDevicePool` will also implement a new interface `DeviceFilterI` so that it can support its own set of `selectors` and filtering algorithms.

### Alternatives Considered
* Create device specific standalone plugins for other SR-IOV devices. Some of the code from this plugin could be re-used as library import in these new plugins.

### Performance Implications
* From device plugin point of view its negligible. 

### Dependencies
* None

### Engineering Impact
*  The binary size, start-up time, build time, test times all expected to be increased.

* Device specific codes will reside in their own package.

### Platforms and Environments
* We are targeting BareMetal use cases only. There is no plan of supporting VM use cases(SR-IOV devices in VMs).

### Compatibility
* These changes should be fully backward compatible with current behaviour and user configurations

### User Impact
* None

**Comment by zshi-redhat** [comment]
Hi @ahalim-intel Thanks for working out the proposal! This is a significant move from current projet scope, I'm sure this'll bring up more challenges and most importantly opptunities ahead.

> # Add support for other SR-IOV devices

> ## Objective
> The objective of this proposal is to extend SR-IOV network device plugin to support other SR-IOV capable devices for example FPGA, Accelerators, SmartNICs etc.

With this definition, I think we're open to any hardware, SR-IOV devices.

> 
> ## Motivation
> 
> ## User Benefit
> 
> ## Design Proposal
> Current implementation of the device plugin focuses on physical SR-IOV networking devices in BareMetal only. The device discovert, resource configurations and selectors are very much network device centric.
> 
> This proposal suggests the following changes in current implementation so that handling other type of devices can be done by implementing some shared interfaces for generic SR-IOV device, not limited to network devices only.

MIght be good to change the repo name to `sriov-device-plugin`?

> 
> 1. Extract a generic `PciDevice` interface from `PciNetDevice`. Then each device type will extend this as shown in the example below for `PciNetDevice`.
> 
> Current `PciNetDevice` interface:
> 
> ```go
> type PciNetDevice interface {
> 	GetPFName() string
> 	GetPfPciAddr() string
> 	GetVendor() string
> 	GetDriver() string
> 	GetDeviceCode() string
> 	GetPciAddr() string
> 	GetNetName() string
> 	IsSriovPF() bool
> 	GetLinkSpeed() string
> 	GetLinkType() string
> 	GetSubClass() string
> 	GetDeviceSpecs() []*pluginapi.DeviceSpec
> 	GetEnvVal() string
> 	GetMounts() []*pluginapi.Mount
> 	GetAPIDevice() *pluginapi.Device
> 	GetRdmaSpec() RdmaSpec
> 	GetVFID() int
> 	GetDDPProfiles() string
> }
> ```
> 
> Change this to:
> 
> ```go
> // PciDevice provides an interface to get device specific information
> type PciDevice interface {
> 	GetVendor() string
> 	GetDriver() string
> 	GetDeviceCode() string
> 	GetPciAddr() string
> 	IsSriovPF() bool
> 	GetSubClass() string
> 	GetDeviceSpecs() []*pluginapi.DeviceSpec
> 	GetEnvVal() string
> 	GetMounts() []*pluginapi.Mount
> 	GetAPIDevice() *pluginapi.Device
> 	GetVFID() int
> 	GetPfPciAddr() string
> }
> 
> // PciNetDevice extends PciDevice
> type PciNetDevice interface {
> 	PciDevice
> 	GetPFName() string
> 	GetNetName() string
> 	GetLinkSpeed() string
> 	GetLinkType() string
> 	GetRdmaSpec() RdmaSpec
>     GetDDPProfiles() string
> }
> ```

Thanks for sorting details out! This clearly illustrates the design idea.

> 
> This will allow the plugin to support other new devices with device specific interface and behaviours. The device specific interface will allow to have device specific `selectors`. For example,
> 
> ```go
> // PciAccelDevice extends PciDevice
> type PciNetDevice interface {
> 	PciDevice
>     SomeFPGAMethod()
> }
> ```
> 
> 1. Introduce a new user defined resourceConfig parameter `deviceType`. This parameter will be optional and default to `netdevice` for backward compatibility.
> 
> ```json
> {
>     "resourceList": [{
>             "resourceName": "intel_sriov_netdevice",
>             "selectors": {
>                 "vendors": ["8086"],
>                 "devices": ["154c", "10ed"],
>                 "drivers": ["i40evf", "ixgbevf"],
>                 "pfNames": ["enp0s0f0","enp2s2f1"]
>             }
>         },
>         {
>             "resourceName": "intel_sriov_fec",
>             "resourcePrefix": "intel.com",
>             "deviceType": "accelerator",
>             "selectors": {
>                 "vendors": ["8086"],
>                 "devices": ["ffff"],
>                 "drivers": ["vfio-pci"],
>                 "other_fec_selector": ["values"]
>             }
>         }
>     ]
> }
> ```
> 

Some fields (e.g. isRdma) defined in current resourceConfig may not be meaningful for other deviceTypes. Shall we do the same change to `resourceConfig` as `pciDevice` to have a inherited structure for each deviceType which contains deviceType specific flags and selectors. And use deviceType to determine which sub-resourceConfig structure to unmarshal data to.

> 1. The SR-IOV plugin manager will keep a map of `PciDevicePool` pool with `deviceType` as key. Each `PciDevicePool` will contain its own list of `PciDevice`. Eeah resource config will create a new kubelet device plugin enpoint serving devices from type of `PciDevicePool`.

Do we need per-resourceServer control? For example, the lifecycle of different device type server may vary, restart/stop of a certain device type server may have impact on other devices as they are exposed via the single SR-IOV device plugin control image/binary.

> 2. The `discoverHostDevices()` will discover devices for all supported `deviceType` using device specific methods.

We might be able to just discover devices specified in resourceConfig deviceType(s).

> 3. Each `PciDevicePool` will also implement a new interface `DeviceFilterI` so that it can support its own set of `selectors` and filtering algorithms.

Another relevant change might be to update the environment variable name exposed via DP API, from fixed prefix `PCIDEVICE` to per device type prefix such as `PCINETDEVICE` for network devices or just use per resource pool `GetEnvs`.

> 
> ### Alternatives Considered
> * Create device specific standalone plugins for other SR-IOV devices. Some of the code from this plugin could be re-used as library import in these new plugins.
> 
> ### Performance Implications
> * From device plugin point of view its negligible.

I think this has reasons that device plugin currently only does reading actions such as discovering, probing, filtering of devices and there is no per device configurtion(writing) being done. However, we have seen attempts to extend device plugin to cover driver binding (of vfio-pci) which might have an impact allocating performance. With more device types supported, I'm not clear if there will be such requests to pre-configure resource before allocating to Pod (will FPGA be such device ?).

> 
> ### Dependencies
> * None
> 
> ### Engineering Impact
> * The binary size, start-up time, build time, test times all expected to be increased.

This may be not obvious when running device plugin standalone, but if it's integrated in a high-level component such as SR-IOV Operator, then we'd care about the start-up time as Operator needs to restart DP several times during device provisioning and the accumulated time might be siginificant.
Let's see how much it will affect the actual start-up time.

For testing, I think we may want to look at 3rd Party CI again to make sure every code change is verified on real hardware. We are doing this testing manually now.

Wrt binary size, I'm a little concerned about image size as I'm not sure if we need to add utility tools for certain device types, such as FPGA device.

I think with current proposal, we will be using one image for all device types. I don't see a particular reason to do one image per device type now, but from integration point of view, one image per device type may work better function wise. For example, we don't need to carry build-in functions or untilities designed for non-network devices in SR-IOV network Operator.

> * Device specific codes will reside in their own package.

I assume this will require some changes in dir structure under `pkg`.

> 
> ### Platforms and Environments
> * We are targeting BareMetal use cases only. There is no plan of supporting VM use cases(SR-IOV devices in VMs).

nit: With more types of devices added, I think it becomes difficult to support VM use cases as different device types may vary on how they be used in VM cases.

> 
> ### Compatibility
> * These changes should be fully backward compatible with current behaviour and user configurations

+2, this is important for other components that integrates DP for device management.

> 
> ### User Impact
> * None

**Comment by amorenoz** [comment]
+1 (thinking about how to add support for vDPA devices)

>> `{
    "resourceList": [{
            "resourceName": "intel_sriov_netdevice",
            "selectors": {
                "vendors": ["8086"],
                "devices": ["154c", "10ed"],
                "drivers": ["i40evf", "ixgbevf"],
                "pfNames": ["enp0s0f0","enp2s2f1"]
            }
        },
        {
            "resourceName": "intel_sriov_fec",
            "resourcePrefix": "intel.com",
            "deviceType": "accelerator",
            "selectors": {
                "vendors": ["8086"],
                "devices": ["ffff"],
                "drivers": ["vfio-pci"],
                "other_fec_selector": ["values"]
            }
        }
    ]
}`


> Some fields (e.g. isRdma) defined in current resourceConfig may not be meaningful for other deviceTypes. Shall we do the same change to resourceConfig as pciDevice to have a inherited structure for each deviceType which contains deviceType specific flags and selectors. And use deviceType to determine which sub-resourceConfig structure to unmarshal data to.

+1 Providing deviceType-specific parameters in resourceConfig would definitely be very useful.

Also, could the deviceType (or even deviceType-specific data) be passed down to other components (CNI, pod, etc) by some means?
Maybe the `Annotations          map[string]string` field that's currently available in the `api.ContainerAllocateResponse` could be used?

**Comment by zshi-redhat** [comment]
> Also, could the deviceType (or even deviceType-specific data) be passed down to other components (CNI, pod, etc) by some means?
> Maybe the `Annotations map[string]string` field that's currently available in the `api.ContainerAllocateResponse` could be used?

Depends on how the deviceType info or data will be used, it can be passed down as environment variable or pod annotation via Device Plugin API. Environment variables can be accessed by application directly inside container. Pod annotation can be mapped to container as a file via downward API.

**Comment by adrianchiris** [comment]
I think that extending SR-IOV network device plugin to support generic SR-IOV devices blurs the boundaries of what this project should and should not cover and will introduce unneeded complexity (as "all in one" projects usually do).

Taking a similar route as CNI seems better to me. That is, defining a new `libsriovdp` library with common code to be shared across multiple SR-IOV CNIs. (This project can be a reference for the structure of future DPs utilizing the new lib)

I see this path is proposed in the alternative section, and i think we should list the a Pro's and Con's for the two and discuss in google doc/meeting.

Thinking of the Operators use-case as Operators today are functionality specific they will all need to run an instance of this "fat" binary with different configurations.(@zshi-redhat  am i right here ?).
which in my opinion is sub-optimal.

**Comment by adrianchiris** [comment]
> 4. The discoverHostDevices() will discover devices for all supported deviceType using device specific methods.
> 
> 5. Each PciDevicePool will also implement a new interface DeviceFilterI so that it can support its own set of selectors and filtering algorithms.

This sounds to me like its likely to be vendor specific (or likely to go that path very fast). i keep thinking of FPGAs, and excuse my ignorance here. 
How would one select an FPGA resource having a specific accelerator logic programmed ?
is there a standard sysfs file for an FPGA device that will expose some sort of image UUID  which is currently programmed onto the FPGA?

**Comment by zshi-redhat** [comment]
> Thinking of the Operators use-case as Operators today are functionality specific they will all need to run an instance of this "fat" binary with different configurations.(@zshi-redhat am i right here ?).
> which in my opinion is sub-optimal.

Yes if we were to use the same image or binary for all device types.
I have a question above about whether this would need per-resource control and per device type image/binary.

**Comment by zshi-redhat** [comment]
> > 1. The discoverHostDevices() will discover devices for all supported deviceType using device specific methods.
> > 2. Each PciDevicePool will also implement a new interface DeviceFilterI so that it can support its own set of selectors and filtering algorithms.
> 
> This sounds to me like its likely to be vendor specific (or likely to go that path very fast). i keep thinking of FPGAs, and excuse my ignorance here.
> How would one select an FPGA resource having a specific accelerator logic programmed ?
> is there a standard sysfs file for an FPGA device that will expose some sort of image UUID which is currently programmed onto the FPGA?
> 
I think above questions need to be answered.

Also will FPGA programming need to be supported in DP?
If yes, I assume FPGA programming will need vendor specific tools, is it possible to have a generic abstraction for support vendor specific programming?

**Comment by zshi-redhat** [comment]
> > 1. The discoverHostDevices() will discover devices for all supported deviceType using device specific methods.
> > 2. Each PciDevicePool will also implement a new interface DeviceFilterI so that it can support its own set of selectors and filtering algorithms.
> 
> This sounds to me like its likely to be vendor specific (or likely to go that path very fast). i keep thinking of FPGAs, and excuse my ignorance here.
> How would one select an FPGA resource having a specific accelerator logic programmed ?
> is there a standard sysfs file for an FPGA device that will expose some sort of image UUID which is currently programmed onto the FPGA?
> 
I think above questions need to be answered.

Also will FPGA programming need to be supported in DP?
If yes, I assume FPGA programming will need vendor specific tools, is it possible to have a generic abstraction for support vendor specific programming?

**Comment by zshi-redhat** [comment]
> > 1. The discoverHostDevices() will discover devices for all supported deviceType using device specific methods.
> > 2. Each PciDevicePool will also implement a new interface DeviceFilterI so that it can support its own set of selectors and filtering algorithms.
> 
> This sounds to me like its likely to be vendor specific (or likely to go that path very fast). i keep thinking of FPGAs, and excuse my ignorance here.
> How would one select an FPGA resource having a specific accelerator logic programmed ?
> is there a standard sysfs file for an FPGA device that will expose some sort of image UUID which is currently programmed onto the FPGA?

I think above questions need to be answered.

Also will FPGA programming need to be supported in DP?
If yes, I assume FPGA programming will need vendor specific tools, is it possible to have a generic abstraction for support vendor specific programming?
I think above questions need to be answered.

Also will FPGA programming need to be supported in DP?
If yes, I assume FPGA programming will need vendor specific tools, is it possible to have a generic abstraction for supporting vendor specific programming?

**Comment by ahalim-intel** [comment]
> Hi @ahalim-intel Thanks for working out the proposal! This is a significant move from current projet scope, I'm sure this'll bring up more challenges and most importantly opptunities ahead.
> 
> > ## Objective
> > The objective of this proposal is to extend SR-IOV network device plugin to support other SR-IOV capable devices for example FPGA, Accelerators, SmartNICs etc.
> 
> With this definition, I think we're open to any hardware, SR-IOV devices.
I think the main motivation of using this plugin for other SR-IOV devices is (1) code re-use, as some of the functionalities are applicable for any SR-IOV capable PCI devices and (2) the handling vfio/uio specific device files. 
> 
> > ## Design Proposal
> > This proposal suggests the following changes in current implementation so that handling other type of devices can be done by implementing some shared interfaces for generic SR-IOV device, not limited to network devices only.
> 
> MIght be good to change the repo name to `sriov-device-plugin`?
Ideally yes. Not sure if we should do this though. It may break downstream projects that are importing some packages from this repo.
 
> 
> > 1. Extract a generic `PciDevice` interface from `PciNetDevice`. Then each device type will extend this as shown in the example below for `PciNetDevice`.
> > 
> > Current `PciNetDevice` interface:
> > Change this to:
> 
> Thanks for sorting details out! This clearly illustrates the design idea.
> 
> > This will allow the plugin to support other new devices with device specific interface and behaviours. The device specific interface will allow to have device specific `selectors`. For example,
> > ```go
> > // PciAccelDevice extends PciDevice
> > type PciNetDevice interface {
> > 	PciDevice
> >     SomeFPGAMethod()
> > }
> > ```
> > 
> > 
> > 
> > 1. Introduce a new user defined resourceConfig parameter `deviceType`. This parameter will be optional and default to `netdevice` for backward compatibility.
> > 
> > ```json
> > {
> >     "resourceList": [{
> >             "resourceName": "intel_sriov_netdevice",
> >             "selectors": {
> >                 "vendors": ["8086"],
> >                 "devices": ["154c", "10ed"],
> >                 "drivers": ["i40evf", "ixgbevf"],
> >                 "pfNames": ["enp0s0f0","enp2s2f1"]
> >             }
> >         },
> >         {
> >             "resourceName": "intel_sriov_fec",
> >             "resourcePrefix": "intel.com",
> >             "deviceType": "accelerator",
> >             "selectors": {
> >                 "vendors": ["8086"],
> >                 "devices": ["ffff"],
> >                 "drivers": ["vfio-pci"],
> >                 "other_fec_selector": ["values"]
> >             }
> >         }
> >     ]
> > }
> > ```
> 
> Some fields (e.g. isRdma) defined in current resourceConfig may not be meaningful for other deviceTypes. Shall we do the same change to `resourceConfig` as `pciDevice` to have a inherited structure for each deviceType which contains deviceType specific flags and selectors. And use deviceType to determine which sub-resourceConfig structure to unmarshal data to.
> 
Yes, that's a good idea. 
> > 1. The SR-IOV plugin manager will keep a map of `PciDevicePool` pool with `deviceType` as key. Each `PciDevicePool` will contain its own list of `PciDevice`. Eeah resource config will create a new kubelet device plugin enpoint serving devices from type of `PciDevicePool`.
> 
> Do we need per-resourceServer control? For example, the lifecycle of different device type server may vary, restart/stop of a certain device type server may have impact on other devices as they are exposed via the single SR-IOV device plugin control image/binary.
> 
Each server is running in its own go thread. I don't see how one can impact another during start-up time though. They will just start at their own time. Yes, in the event of stopping/killing the plugin's main process, it will wait until all servers are stopped before main exits.

> > 1. The `discoverHostDevices()` will discover devices for all supported `deviceType` using device specific methods.
> 
> We might be able to just discover devices specified in resourceConfig deviceType(s).
Yes, but in many cases we may do the same discovery multiple times, depending on how many resourceConfig with same deviceTypes. I think it is time saving to enumerate a initial global device lists for each deviceType and then only apply selectors as per resourceConfig.
> 
> > 1. Each `PciDevicePool` will also implement a new interface `DeviceFilterI` so that it can support its own set of `selectors` and filtering algorithms.
> 
> Another relevant change might be to update the environment variable name exposed via DP API, from fixed prefix `PCIDEVICE` to per device type prefix such as `PCINETDEVICE` for network devices or just use per resource pool `GetEnvs`.
> 
Will this be necessary? Yes, `GetEnvs` is there should we want to implement device specific ENV naming convention. I'd prefer not to change existing naming convention that we are using now for network devices. This will avoid unnecessary compatibility issue. The `PCIDEVICE` prefix for env is generic already that can go with all deviceType. 

> > ### Alternatives Considered
> > 
> > * Create device specific standalone plugins for other SR-IOV devices. Some of the code from this plugin could be re-used as library import in these new plugins.
> > 
> > ### Performance Implications
> > 
> > * From device plugin point of view its negligible.
> 
> I think this has reasons that device plugin currently only does reading actions such as discovering, probing, filtering of devices and there is no per device configurtion(writing) being done. However, we have seen attempts to extend device plugin to cover driver binding (of vfio-pci) which might have an impact allocating performance. With more device types supported, I'm not clear if there will be such requests to pre-configure resource before allocating to Pod (will FPGA be such device ?).
> 
We will try to steer clear from runtime device changes with device plugin. First of all, the device plugin API does not provide any runtime config to the plugin, so I don't see any reason why we should treat request 'a' any differently from requests 'b' other than the deviceID. For FPGA use cases, I think the device need to be pre-programmed before it's enumerating its VFs and registering them to Kubelet with device plugin. Same as we do now with SR-IOV network devices - we create VFs, bind drivers then launch the plugin to register them with Kubelet.
> > ### Dependencies
> > 
> > * None
> > 
> > ### Engineering Impact
> > 
> > * The binary size, start-up time, build time, test times all expected to be increased.
> 
> This may be not obvious when running device plugin standalone, but if it's integrated in a high-level component such as SR-IOV Operator, then we'd care about the start-up time as Operator needs to restart DP several times during device provisioning and the accumulated time might be siginificant.
> Let's see how much it will affect the actual start-up time.
> 
> For testing, I think we may want to look at 3rd Party CI again to make sure every code change is verified on real hardware. We are doing this testing manually now.
> 
> Wrt binary size, I'm a little concerned about image size as I'm not sure if we need to add utility tools for certain device types, such as FPGA device.
> 
> I think with current proposal, we will be using one image for all device types. I don't see a particular reason to do one image per device type now, but from integration point of view, one image per device type may work better function wise. For example, we don't need to carry build-in functions or untilities designed for non-network devices in SR-IOV network Operator.
> 
Are we talking about creating separate binary for each deviceType?

> > * Device specific codes will reside in their own package.
> 
> I assume this will require some changes in dir structure under `pkg`.
> 
> > ### Platforms and Environments
> > 
> > * We are targeting BareMetal use cases only. There is no plan of supporting VM use cases(SR-IOV devices in VMs).
> 
> nit: With more types of devices added, I think it becomes difficult to support VM use cases as different device types may vary on how they be used in VM cases.
> 
> > ### Compatibility
> > 
> > * These changes should be fully backward compatible with current behaviour and user configurations
> 
> +2, this is important for other components that integrates DP for device management.
> 
> > ### User Impact
> > 
> > * None

**Comment by ahalim-intel** [comment]
> > > 1. The discoverHostDevices() will discover devices for all supported deviceType using device specific methods.
> > > 2. Each PciDevicePool will also implement a new interface DeviceFilterI so that it can support its own set of selectors and filtering algorithms.
> > 
> > 
> > This sounds to me like its likely to be vendor specific (or likely to go that path very fast). i keep thinking of FPGAs, and excuse my ignorance here.
> > How would one select an FPGA resource having a specific accelerator logic programmed ?
> > is there a standard sysfs file for an FPGA device that will expose some sort of image UUID which is currently programmed onto the FPGA?
> 
> I think above questions need to be answered.
> 
> Also will FPGA programming need to be supported in DP?
> If yes, I assume FPGA programming will need vendor specific tools, is it possible to have a generic abstraction for support vendor specific programming?

As replied to other comment above, the device plugin will not be concerned with programming/re-programming an FPGA device. With my limited knowledge on FPGA, my understanding is one can only program the entire device not a VF(we are talking about FPGA in SR-IOV mode here, an FPGA can also be used without SR-IOV and there's other plugin for that already). So, it doesn't make sense to me to program an FPGA per request for VFs. The device programming should be done before launching the device plugin. However, an FPGA may provide some mechanism/interface to query about program information that could be useful in the FPGA selector - should a resourcePool targets a specifically programmed FPGA.

**Comment by amorenoz** [comment]
> 
> > > ### Alternatives Considered
> > > 
> > > * Create device specific standalone plugins for other SR-IOV devices. Some of the code from this plugin could be re-used as library import in these new plugins.
> > > 
> > > ### Performance Implications
> > > 
> > > * From device plugin point of view its negligible.
> > 
> > 
> > I think this has reasons that device plugin currently only does reading actions such as discovering, probing, filtering of devices and there is no per device configurtion(writing) being done. However, we have seen attempts to extend device plugin to cover driver binding (of vfio-pci) which might have an impact allocating performance. With more device types supported, I'm not clear if there will be such requests to pre-configure resource before allocating to Pod (will FPGA be such device ?).
> 
> We will try to steer clear from runtime device changes with device plugin. First of all, the device plugin API does not provide any runtime config to the plugin, so I don't see any reason why we should treat request 'a' any differently from requests 'b' other than the deviceID. For FPGA use cases, I think the device need to be pre-programmed before it's enumerating its VFs and registering them to Kubelet with device plugin. Same as we do now with SR-IOV network devices - we create VFs, bind drivers then launch the plugin to register them with Kubelet.

Would you reject having those runtime actions deviceType-specific so they don't affect other deviceTypes? I'm thinking about a device being shared between two pools and consumed differently depending on the pool (i.e: the deviceType) it gets allocated through. Does that make any sense?

**Comment by zshi-redhat** [comment]
> > 
> > Do we need per-resourceServer control? For example, the lifecycle of different device type server may vary, restart/stop of a certain device type server may have impact on other devices as they are exposed via the single SR-IOV device plugin control image/binary.
> 
> Each server is running in its own go thread. I don't see how one can impact another during start-up time though. They will just start at their own time. Yes, in the event of stopping/killing the plugin's main process, it will wait until all servers are stopped before main exits.

From [here](https://github.com/open-ness/specs/blob/master/doc/enhanced-platform-awareness/openness-fpga.md#intel-pac-n3000-remote-system-update-flow-in-openness-network-edge-kubernetes), It looks like FPGA VFs and queues will be provisioned by FPGA program separately. If we were to have SR-IOV Operator manage one SR-IOV DP for both FPGA VF and network VF,  will that require per resourceServer control? for example, restarting FPGA resource server to re-discover FPGA device change or restart network resource server when there is a change on numVFs of network device confgured via Operator node policy.

> 
> > > 1. The `discoverHostDevices()` will discover devices for all supported `deviceType` using device specific methods.
> > 
> > 
> > We might be able to just discover devices specified in resourceConfig deviceType(s).
> > Yes, but in many cases we may do the same discovery multiple times, depending on how many resourceConfig with same deviceTypes. I think it is time saving to enumerate a initial global device lists for each deviceType and then only apply selectors as per resourceConfig.
> > > 1. Each `PciDevicePool` will also implement a new interface `DeviceFilterI` so that it can support its own set of `selectors` and filtering algorithms.
> > 
> > 
> > Another relevant change might be to update the environment variable name exposed via DP API, from fixed prefix `PCIDEVICE` to per device type prefix such as `PCINETDEVICE` for network devices or just use per resource pool `GetEnvs`.
> 
> Will this be necessary? Yes, `GetEnvs` is there should we want to implement device specific ENV naming convention. I'd prefer not to change existing naming convention that we are using now for network devices. This will avoid unnecessary compatibility issue. The `PCIDEVICE` prefix for env is generic already that can go with all deviceType.

I think the idea of having PCIDEVICE_* env is to help application to get device PCI address.

How would FPGA VF be presented inside container, will it look the same as a kernel network VF device?  I'm thinking how application would easily get the identifier of device if there are multiple types of VFs allocated for one container. WIth PCIDEVICE_*,  application cannot tell which one is FPGA device or which one is network device. Shall we encouge user to detect deviceType by getting device driver or reading resource name from environment variable?


> 
> > > ### Alternatives Considered
> > > 
> > > * Create device specific standalone plugins for other SR-IOV devices. Some of the code from this plugin could be re-used as library import in these new plugins.
> > > 
> > > ### Performance Implications
> > > 
> > > * From device plugin point of view its negligible.
> > 
> > 
> > I think this has reasons that device plugin currently only does reading actions such as discovering, probing, filtering of devices and there is no per device configurtion(writing) being done. However, we have seen attempts to extend device plugin to cover driver binding (of vfio-pci) which might have an impact allocating performance. With more device types supported, I'm not clear if there will be such requests to pre-configure resource before allocating to Pod (will FPGA be such device ?).
> 
> We will try to steer clear from runtime device changes with device plugin. First of all, the device plugin API does not provide any runtime config to the plugin, so I don't see any reason why we should treat request 'a' any differently from requests 'b' other than the deviceID. For FPGA use cases, I think the device need to be pre-programmed before it's enumerating its VFs and registering them to Kubelet with device plugin. Same as we do now with SR-IOV network devices - we create VFs, bind drivers then launch the plugin to register them with Kubelet.

Yes, DP API doesn't support any runtime config, but it's possible that those configs are defined in DP configmap and be applied during DP start time or allocating time. My concern is will this config path be used later by FPGA programming?  It might be better to make it clear in the proposal.

> 
> > > ### Dependencies
> > > 
> > > * None
> > > 
> > > ### Engineering Impact
> > > 
> > > * The binary size, start-up time, build time, test times all expected to be increased.
> > 
> > 
> > This may be not obvious when running device plugin standalone, but if it's integrated in a high-level component such as SR-IOV Operator, then we'd care about the start-up time as Operator needs to restart DP several times during device provisioning and the accumulated time might be siginificant.
> > Let's see how much it will affect the actual start-up time.
> > For testing, I think we may want to look at 3rd Party CI again to make sure every code change is verified on real hardware. We are doing this testing manually now.
> > Wrt binary size, I'm a little concerned about image size as I'm not sure if we need to add utility tools for certain device types, such as FPGA device.
> > I think with current proposal, we will be using one image for all device types. I don't see a particular reason to do one image per device type now, but from integration point of view, one image per device type may work better function wise. For example, we don't need to carry build-in functions or untilities designed for non-network devices in SR-IOV network Operator.
> 
> Are we talking about creating separate binary for each deviceType?
> 

It's about separate image for each deviceType. As I don't know if all the device with different types will be managed by single component(like SR-IOV Operator) in a cluster that launches one DP daemon. It's possible that there will be several DP daemons running in a cluster, with each for a different device type.

**Comment by damiankopyto** [comment]
> > > > 1. The discoverHostDevices() will discover devices for all supported deviceType using device specific methods.
> > > > 2. Each PciDevicePool will also implement a new interface DeviceFilterI so that it can support its own set of selectors and filtering algorithms.
> > > 
> > > 
> > > This sounds to me like its likely to be vendor specific (or likely to go that path very fast). i keep thinking of FPGAs, and excuse my ignorance here.
> > > How would one select an FPGA resource having a specific accelerator logic programmed ?
> > > is there a standard sysfs file for an FPGA device that will expose some sort of image UUID which is currently programmed onto the FPGA?
> > 
> > 
> > I think above questions need to be answered.
> > Also will FPGA programming need to be supported in DP?
> > If yes, I assume FPGA programming will need vendor specific tools, is it possible to have a generic abstraction for support vendor specific programming?
> 
> As replied to other comment above, the device plugin will not be concerned with programming/re-programming an FPGA device. With my limited knowledge on FPGA, my understanding is one can only program the entire device not a VF(we are talking about FPGA in SR-IOV mode here, an FPGA can also be used without SR-IOV and there's other plugin for that already). So, it doesn't make sense to me to program an FPGA per request for VFs. The device programming should be done before launching the device plugin. However, an FPGA may provide some mechanism/interface to query about program information that could be useful in the FPGA selector - should a resourcePool targets a specifically programmed FPGA.

Hi all, let me give some more insight and clarification on some of these queries:
1. FPGA programming would not be supported from DP, the FPGA would be programmed by tools such as OPAE (it could be even programmed through JTAG using tools like Quartus, or in case of other vendors via other mechanism), it would be independent of DP and assumption is made that the FPGA used is programmed/configured before being used by DP (You can think of it similar to NIC, it is plugged into the platform with its hardware functionality pre-configured, there is no expectation on DP to alter this functionality). Similarly the DP would not be responsible for binding the devices to appropriate driver and creating the VFs, the responsibility is on the user to do it as part of platform preparation.

2. This would not be vendor specific, the DP does not really need to concern itself with what sort of FPGA program is loaded - as long as the programmed FPGA exposes SRIOV devices that are bound to appropriate  driver, they would be requested from ConfigMap and device specific files would be allocated to pod.

3.  I think that the mechanism to determine the FPGA program loaded would be vendor specific, from my experience this would be done through OPAE, this would not necessarily be needed as the user should be aware of what the FPGA is programmed with. Additionally if the FPGA was programmed for different functionality then most likely it should be discovered as a different device with different DeviceID (ie. Functionality 'A' = devId "1000", Functionality 'B' = devId "1001"), the only benefit here would be differentiating between different versions of same loaded program - but this is not crucial.

**Comment by pliurh** [comment]
I think we need to have a consensus on the scope of the devices shall be managed by this plugin first. Do we really want to have one DP to "support other SR-IOV capable devices for example FPGA, Accelerators, SmartNICs etc"? The scope looks too wide to me. It's hard to have a universal way to discover and let the pod consume those devices. 

Also, lets us think about this from an operational perspective. For each kind of abovementioned devices, users may need to initialize/program it in different ways before consuming it by K8S. Now for the network device, we have the SRIOV network operator to take care of those things in one place. However, it is not realistic to have one controller/operator to manage all those devices with the fact that they are so different. And I don't think you want to have multiple controllers which manage one device plugin either.

I agree with @adrianchiris that "defining a new libsriovdp library with common code" would be a better approach.

**Comment by ahalim-intel** [comment]
@pliurh Yes, that is the intention of this proposal is to get others opinion, input and concern or maybe some alternative proposal.
I agree this will increase the scope of this plugin. But as you know this plugin already able to create multiple resource endpoints plus it also can discover other PCI devices already, even though we only looking at network devices. So this extension really is about allowing those devices in the supported device list with their own selectors.
This will allow users to just add additional set of configuration to expose those devices. This plugin takes care of the grpc server lifecycle, handling driver specific device files, grouping devices into resource pools etc. With library approach other plugins will have to re-implement these functionality again at some extent.
 
Regarding  SRIOV network operator, in my opinion this operator should only be concerned with SR-IOV networking as it does now. This extension should not break it in anyway. Users should not expect to use this operator for non-networking use-cases. If they need similar operator to manage the FPGA devices(or other SR-IOV devices) for handling the re-programming etc. it's obvious they will need a different operator for that because SRIOV CNI, net-attach-def etc. won't make sense in non-networking scenarios.

**Comment by krishnajs** [comment]
> > > > > 1. The discoverHostDevices() will discover devices for all supported deviceType using device specific methods.
> > > > > 2. Each PciDevicePool will also implement a new interface DeviceFilterI so that it can support its own set of selectors and filtering algorithms.
> > > > 
> > > > 
> > > > This sounds to me like its likely to be vendor specific (or likely to go that path very fast). i keep thinking of FPGAs, and excuse my ignorance here.
> > > > How would one select an FPGA resource having a specific accelerator logic programmed ?
> > > > is there a standard sysfs file for an FPGA device that will expose some sort of image UUID which is currently programmed onto the FPGA?
> > > 
> > > 
> > > I think above questions need to be answered.
> > > Also will FPGA programming need to be supported in DP?
> > > If yes, I assume FPGA programming will need vendor specific tools, is it possible to have a generic abstraction for support vendor specific programming?
> > 
> > 
> > As replied to other comment above, the device plugin will not be concerned with programming/re-programming an FPGA device. With my limited knowledge on FPGA, my understanding is one can only program the entire device not a VF(we are talking about FPGA in SR-IOV mode here, an FPGA can also be used without SR-IOV and there's other plugin for that already). So, it doesn't make sense to me to program an FPGA per request for VFs. The device programming should be done before launching the device plugin. However, an FPGA may provide some mechanism/interface to query about program information that could be useful in the FPGA selector - should a resourcePool targets a specifically programmed FPGA.
> 
> Hi all, let me give some more insight and clarification on some of these queries:
> 
> 1. FPGA programming would not be supported from DP, the FPGA would be programmed by tools such as OPAE (it could be even programmed through JTAG using tools like Quartus, or in case of other vendors via other mechanism), it would be independent of DP and assumption is made that the FPGA used is programmed/configured before being used by DP (You can think of it similar to NIC, it is plugged into the platform with its hardware functionality pre-configured, there is no expectation on DP to alter this functionality). Similarly the DP would not be responsible for binding the devices to appropriate driver and creating the VFs, the responsibility is on the user to do it as part of platform preparation.
> 2. This would not be vendor specific, the DP does not really need to concern itself with what sort of FPGA program is loaded - as long as the programmed FPGA exposes SRIOV devices that are bound to appropriate  driver, they would be requested from ConfigMap and device specific files would be allocated to pod.
> 3. I think that the mechanism to determine the FPGA program loaded would be vendor specific, from my experience this would be done through OPAE, this would not necessarily be needed as the user should be aware of what the FPGA is programmed with. Additionally if the FPGA was programmed for different functionality then most likely it should be discovered as a different device with different DeviceID (ie. Functionality 'A' = devId "1000", Functionality 'B' = devId "1001"), the only benefit here would be differentiating between different versions of same loaded program - but this is not crucial.

+, in general, the FPGA programming interface when exposed through PCIe will be handled at Physical Function level. The programming stack in most cases would need privileges. This SRIOV device plugin is not meant to address the FPGA programming interface.

**Comment by Levovar** [comment]
if you are worried about scope-creep let's not include the Operator -which is not part of this project :) - into the discussion.
I'm generally fine with a library approach as well, but IMO the existing DP is basically already capable of handling multiple device types. FPGA devices are not the first device type requiring specific discovery and mounting, vfio or igb_uio bound devices are the same from this perspective.
So I generally agree that we only need the right abstraction to capture this already existing requirment, and because of that I personally like the proposal.

I also like it because it would re-open the possibility of using the DP to handle PFs -for exclusive 1:1 Pod-NIC DPDK acecss- both in VM, or in a bare metal environment; a feature originally supported, but later forgotten.
Possibly it would be an unintended side-effect :)

**Comment by adrianchiris** [comment]
@damiankopyto 
> This would not be vendor specific, the DP does not really need to concern itself with what sort of FPGA program is loaded - as long as the programmed FPGA exposes SRIOV devices that are bound to appropriate driver, they would be requested from ConfigMap and device specific files would be allocated to pod.

Thanks for clarifying! 
So, for example, say a gzip programmed FPGA would expose VFs which will be bound to a specific gzip pci driver which will be used as a selector for the resource in device plugin, am i correct ?

Now, The list of mounts required for a user space application to utilize this device would then need to be specified in a configuration file ? are the paths fixed or  some driver specific magic is required to determine the mounts  ?


@Levovar 
> if you are worried about scope-creep let's not include the Operator -which is not part of this project :) - into the discussion.

I do think that possible constellations where the device plugin is used need to be taken into consideration.

e.g for operators maybe it will require to split to a low-level device plugin operator which will apply sriov device plugin configurations from other operators like sriov network operator.
or just run multiple instances of device plugin, each with its own configuration.

**Comment by ahalim-intel** [comment]
> @damiankopyto
> 
> > This would not be vendor specific, the DP does not really need to concern itself with what sort of FPGA program is loaded - as long as the programmed FPGA exposes SRIOV devices that are bound to appropriate driver, they would be requested from ConfigMap and device specific files would be allocated to pod.
> 
> Thanks for clarifying!
> So, for example, say a gzip programmed FPGA would expose VFs which will be bound to a specific gzip pci driver which will be used as a selector for the resource in device plugin, am i correct ?
> 

> Now, The list of mounts required for a user space application to utilize this device would then need to be specified in a configuration file ? are the paths fixed or some driver specific magic is required to determine the mounts ?
>

Let's try not dig so deep at this stage(and focusing on the FPGA devices only). The other implementation did not concern about FPGA program specific details or how it should be handled. I believe, this proposed changes will leave enough room to tackle the FPGA or other SR-IOV device specific needs in the future. Our goal here now is to refactor the design with some abstraction so that we can accommodate those devices as well without impacting existing networking devices. 
> @Levovar
> 
> > if you are worried about scope-creep let's not include the Operator -which is not part of this project :) - into the discussion.
> 
> I do think that possible constellations where the device plugin is used need to be taken into consideration.
> 
> e.g for operators maybe it will require to split to a low-level device plugin operator which will apply sriov device plugin configurations from other operators like sriov network operator.
> or just run multiple instances of device plugin, each with its own configuration.

**Comment by adrianchiris** [comment]
> Our goal here now is to refactor the design with some abstraction so that we can accommodate those devices as well without impacting existing networking devices.

I think the FPGA is an interesting use-case, a plugin mechanism may be required to get the required mounts/device spec to a container, or keep that logic in project for now.

**Comment by ahalim-intel** [comment]
> > Our goal here now is to refactor the design with some abstraction so that we can accommodate those devices as well without impacting existing networking devices.
> 
> I think the FPGA is an interesting use-case, a plugin mechanism may be required to get the required mounts/device spec to a container, or keep that logic in project for now.

We already have this mechanism with these interfaces:
```
	GetDeviceSpecs() []*pluginapi.DeviceSpec
	GetEnvVal() string
	GetMounts() []*pluginapi.Mount
	GetAPIDevice() *pluginapi.Device
```
A new deviceType can implement any of these interfaces without affecting other devices.

**Comment by zshi-redhat** [comment]
> if you are worried about scope-creep let's not include the Operator -which is not part of this project :) - into the discussion.

I actually consider integration is a big part of this change because DP is likely be used together with other components in a real kube production environment. Looking at GPU device plugin, it also gets integrated with a special resource operator.  Bringing up a DP daemon manually may be more suitable in a test environment in my opinion. Operator is one of those examples of doing this integration work, it could also be other framework or component that doing this integration.

> I'm generally fine with a library approach as well, but IMO the existing DP is basically already capable of handling multiple device types. FPGA devices are not the first devices requiring specific discovery and mounting, vfio or igb_uio bound devices are the same from this perspective.
> So I generally agree that we only need the right abstraction to capture this already existing requirement, and because of that I personally like the proposal.
> 
> I also like it because it would re-open the possibility of using the DP to directly handle PFs -for exclusive 1:1 Pod:phys NIC DPDK allocations- both in VM, or in a bare metal environment; a feature originally supported, but later forgotten.
> Tho I guess this is an unintended side-effect :)

PF support was enabled unintentionally before as side-effect with `rootDevices` in v2.1.0, but it doesn't last long as we moved to selector based filters in later release. side-effect may go away if we didn't explicitly call it out: -)

**Comment by zshi-redhat** [comment]
> 
> Hi all, let me give some more insight and clarification on some of these queries:
> 
Thanks for providing the insight!

> 1. FPGA programming would not be supported from DP, the FPGA would be programmed by tools such as OPAE (it could be even programmed through JTAG using tools like Quartus, or in case of other vendors via other mechanism), it would be independent of DP and assumption is made that the FPGA used is programmed/configured before being used by DP (You can think of it similar to NIC, it is plugged into the platform with its hardware functionality pre-configured, there is no expectation on DP to alter this functionality). Similarly the DP would not be responsible for binding the devices to appropriate driver and creating the VFs, the responsibility is on the user to do it as part of platform preparation.

Ack, FPGA programming is important if we want to have a whole FPGA solution on kubernetes, but it's not necessrily be part of DP. Thanks for clarifying on this as I think there are some confusion and hesitation about supporting vendor specific programming mechensim or toolings in DP.

> 2. This would not be vendor specific, the DP does not really need to concern itself with what sort of FPGA program is loaded - as long as the programmed FPGA exposes SRIOV devices that are bound to appropriate  driver, they would be requested from ConfigMap and device specific files would be allocated to pod.

I understand one factor of re-using DP to support FPGA is that it already supports mounting device spec for interface binded with userspace driver.  Is it possible that FPGA device is not using userspace driver? if yes, what other drivers will be used? Is there a general way of detecting device driver when userspace driver is not used ? 

> 3. I think that the mechanism to determine the FPGA program loaded would be vendor specific, from my experience this would be done through OPAE, this would not necessarily be needed as the user should be aware of what the FPGA is programmed with. Additionally if the FPGA was programmed for different functionality then most likely it should be discovered as a different device with different DeviceID (ie. Functionality 'A' = devId "1000", Functionality 'B' = devId "1001"), the only benefit here would be differentiating between different versions of same loaded program - but this is not crucial.

It may matter if user application needs to know what program is loaded on FPGA device. Would the program info be needed for any reason? (I don't think we want to pack the OPAE into DP image)

**Comment by Levovar** [comment]
> It may matter if user application needs to know what program is loaded on FPGA device. Would the program info be needed for any reason? (I don't think we want to pack the OPAE into DP image)

true, but utlimately it is the operator who configures the FPGA **and** the DP. he knows what program is running on the device, and can easily name the DevicePool accordingly e.g. "smartnic.io/crypto" and "smartnic.io/ovs-offload".
Name of the pool is transparent to the DP, the program is identifiable by the consumer and there is still no dependency between FPGA config and DP.

**Comment by Levovar** [comment]
> PF support was enabled unintentionally before as side-effect with `rootDevices` in v2.1.0, but it doesn't last long as we moved to selector based filters in later release. side-effect may go away if we didn't explicitly call it out: -)

yeah I know; just wanted to call out that the "side-effect" would be welcomed back with open arms ;)
the need to use PFs as scheduleable net devices still exists, and yes, I know one could use the host device CNI but for hte purpose but that approach comes with all the same side-effects solved by the DP (resource scheduling, access control etc.)

**Comment by zshi-redhat** [comment]
> > It may matter if user application needs to know what program is loaded on FPGA device. Would the program info be needed for any reason? (I don't think we want to pack the OPAE into DP image)
> 
> true, but utlimately it is the operator who configures both the FPGA **and** the DP. she knows what program is running on the device, and can easily name the DevicePool accordingly e.g. "smartnic.io/crypto" and "smartnic.io/fec".
> Name of the pool is transparent to the DP, the service provided by these "VFs" is easily identifiable by the consumer and there is still no dependency between FPGA config and DP.
> I personally see the reservation / advertisement of special infrastructure labels and resource names as an upward trend, this would neatly fall into the same category without adding additional dependencies to the DP project

The resource name is identiable by Operator, DP doesn't know the meaning of resource name, it relies on selectors to group a set of device into resource pool by discovering various properties of device. Will there be any need to group devices according to program loaded in FPGA PF device (for example, if there are two FPGA device loaded with same program, Operator wants to group VFs from the two device together)? If so, which tool will be used by DP to discover the loaded program? 

Would it also be a reasonable case that a container image contains application that is capable of using devices with different programs loaded? How would the application know which image is loaded so that it can run different code path or logic? DP has API to pass additional environment variable or annotation to pod, should the program info be passed to container via API?

**Comment by Levovar** [comment]
when I use the word operator with lower case I mean the human, who owns the infra :)

what I tried to highlight is that IMO the DP doesn't need to know what's on the FPGA, and it is alright!
it just needs to provide an API through which it can be told what information to advertise, and the existing selector-based mechanism is perfect for that: "devices on this PF bound to that driver are called FEC, now please go and advertise them"

hence my original comment on the scope-creep: you can definitely try to create an all-knowing DP, but IMO that will lead to a nightmare.
or, provide a generic solution to handle PCI devices the same way, without worrying about the vendor specific aspects
leave it to the vendors. let them create their own Ansible playbooks, Operators (now upper case), Controllers, Informers, systemd services, CLIs etc. to talk to the DP

the existing environment variable pushed into the container is also suitable to identify which PCI device belongs to which service, if the devices are again grouped based on their role / program
"SMARTNIC_IO_FEC=PCI_ID1,PCI_ID2
SMARTNIC_IO_CRYPTO=PCI_ID3"

my two cents :)

**Comment by zshi-redhat** [comment]
> when I use the word operator with lower case I mean the human, who owns the infra :)
> 
> what I tried to highlight is that IMO the DP doesn't need to know what's on the FPGA, and it is alright!

Yes, I won't doubt DP will work without knowing program on FPGA device for some use cases.
The device would have other properties that can be detected by DP without adding device specific knowlege/tools. 

> it just needs to provide an API through which it can be told what information to advertise, and the existing selector-based mechanism is perfect for that: "devices on this PF bound to that driver are called FEC, now please go and advertise them"
> 
The is perfect for user who uses PF selector to group a set of device.
The argument is would that be other cases that user wants to group a set of FPGA device via loaded program? I think we want a design to have potential to support other (if not all) selecting requirement, without sacrificing usability of other device types(such as increase the image size by adding device type specific toolings, lose the granularity of per-device-type resource server control etc).

> hence my original comment on the scope-creep: you can definitely try to create an all-knowing DP, but IMO that will lead to a nightmare.

I agree we cannot support all selecting requirements, but having some guidlines will help us evaluate futher request properly.

> or, provide a generic solution to handle PCI devices the same way, without worrying about the vendor specific aspects
> leave it to the vendors. let them create their own Ansible playbooks, Operators (now upper case), Controllers, Informers, systemd services, CLIs etc. to talk to the DP
> 
> the existing environment variable pushed into the container is also suitable to identify which PCI device belongs to which service, if the devices are again grouped based on their role / program
> "SMARTNIC_IO_FEC=PCI_ID1,PCI_ID2
> SMARTNIC_IO_CRYPTO=PCI_ID3"
> 
Right, the environment variables could be used for application to know which type of resource is allocated to pod. I also suggested in above comments to update the prefix per device type.
But this comes after DP has successfully discovered and grouped devices. The meaning of resource name or environment var is associated with DP selectors.

> my two cents :)

**Comment by killianmuldoon** [comment]
@ahalim-intel This is closed by #214  , right?

**Comment by ahalim-intel** [comment]
@killianmuldoon Yes, this should've been closed.
